subprojects/vinumc/eval.c

```diff
@@ -257,8 +257,38 @@ DO_CALLS_FUNC_SIGNATURE(do_calls_call) {
 			find_symbol_on_scopes(&ctx->scopes, curr_scope, symbol_node->text);
 
 		// expose context to external function
-		struct _call_ctx cctx = { .text = tmp_out.base };
+		struct _call_ctx cctx = { .text = tmp_out.base, .status = false };
+
 		struct return_value call_return = symbol_info->as.func(&cctx);
+		if (call_return.status == false) {
+			size_t node;
+			const struct ast_node *args_node = &VUT_VEC_AT(&ast->nodes, args_id);
+			size_t len = cctx.args_req.len;
+			for (size_t i = 0; i < len; i++) {
+				node = VUT_VEC_AT(&args_node->childs, VUT_VEC_POP(&cctx.args_req));
+				const struct ast_node *arg_node = &VUT_VEC_AT(&ast->nodes, node);
+				if (arg_node->type == ASSIGNMENT) {
+					node = VUT_VEC_AT(&arg_node->childs, 1);
+				}
+				struct vut_str arg_out = {};
+				VUT_VEC_PUT(&arg_out, '\0');
+				do_calls(ctx, ast, &arg_out, node, flags);
+				VUT_VEC_PUT(&cctx.args, arg_out.base);
+			}
+			len = cctx.names_req.len;
+			for (size_t i = 0; i < len; i++) {
+				struct namespace_entry symbol = *find_symbol_on_scopes(
+					&ctx->scopes, curr_scope, VUT_VEC_POP(&cctx.names_req));
+				node = symbol.as.ast_node_id;
+				struct vut_str name_out = {};
+				VUT_VEC_PUT(&name_out, '\0');
+				do_calls(ctx, ast, &name_out, node, flags);
+				VUT_VEC_PUT(&cctx.names, name_out.base);
+			}
+
+			cctx.status = true;
+			call_return = symbol_info->as.func(&cctx);
```

Comment:
We can't call the the function twice on a single vinum call.

What if the external function is not [idempotent](https://en.wikipedia.org/wiki/Idempotence)? For example, a function that adds a single entry to a database: If we call it on vinum code, two entries could be added instead of just a single one.

Is it possible to pre-compute the args before calling the external function?

Reply:
> Is it possible to pre-compute the args before calling the external function?

This is the biggest change, so I'll focus on it first. Precomputing the arguments is possible, which is how I did it before, but for example, to search for a name in the namespace, we would have to provide the scope and the function to search for symbols.
The current method allows for flexibility, without having to provide the internal functions to the external library. To ensure it's idempotent, I thought about requiring the external library to check if ctx.status is true before performing the operations.
I would like to discuss this thoroughly before making any changes, since it alters the entire implementation.
In any case, suggestions are welcome.